### ISSUE TITLE
StackOverflowError when analyzing project

### ISSUE THREAD

**Issue by LubomirVarga** [jrudolph/sbt-dependency-graph#77]
I have got an StackOverflowError exception when analyzing one project with submodules. Possible circular dependencies. Here is stack at the end:

```
        at net.virtualvoid.sbt.graph.Graph$.net$virtualvoid$sbt$graph$Graph$$toAsciiLines$1(Graph.scala:37)
        at net.virtualvoid.sbt.graph.Graph$$anonfun$1.apply(Graph.scala:37)
        at net.virtualvoid.sbt.graph.Graph$$anonfun$1.apply(Graph.scala:37)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
        at scala.collection.Iterator$class.foreach(Iterator.scala:727)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
        at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
        at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
        at scala.collection.AbstractTraversable.map(Traversable.scala:105)
        at net.virtualvoid.sbt.graph.Graph$.net$virtualvoid$sbt$graph$Graph$$toAsciiLines$1(Graph.scala:37)
        at net.virtualvoid.sbt.graph.Graph$$anonfun$1.apply(Graph.scala:37)
        at net.virtualvoid.sbt.graph.Graph$$anonfun$1.apply(Graph.scala:37)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
        at scala.collection.Iterator$class.foreach(Iterator.scala:727)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
        at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
```

**Comment by backuitist** [comment]
This was already reported and a PR was submitted (but discarded)... 
https://github.com/jrudolph/sbt-dependency-graph/pull/40
This boils down to the fact that `Graph.toAsciiLine` isn't tail-recursive, so for large projects (this is uncommon with SBT) it explodes.

**Comment by jrudolph** [comment]
Thanks for the report, @LubomirVarga. I agree that this should be fixed.

@backuitist I cannot remember that you opened a PR, was that here or for sbt itself?

**Comment by jrudolph** [comment]
Thinking about it I think it is very unlikely that this is because of the recursive function. With a normal sized stack it should be possible to have graphs several thousand layers deep. So, this is more likely a circular dependency. 

A reproducer would be much appreciated.

**Comment by backuitist** [comment]
The PR was deleted by the github staff on my demand (I pushed something on the wrong repo by mistake which ended-up in the PR... :s).
Anyhow, I'm fairly certain that `Graph.toAsciiLine` is the culprit. In the commit above - which fixes the bug - i've added a test that demonstrate the problem.
I really don't have time for more so feel free to copy paste my code if that helps :)

**Comment by backuitist** [comment]
The test is called `notStackOverflow`

**Comment by jrudolph** [comment]
Thanks for the explanations, @backuitist. My point was that graphs so deep are unlikely in the context of this plugin, so the fix which probably helped here was your addition of cycle detection.

**Comment by backuitist** [comment]
Ok, now that you insist there's also this test: `cutOffCycles`, so I guess you're right ;)

**Comment by jrudolph** [comment]
:)

**Comment by jrudolph** [comment]
I released a new version 0.8.0-beta1 (to sonatype, wait a few hours to let it appear on maven central) which should have fixed this issue.

@LubomirVarga, could you check if this has fixed your problem?

**Comment by LubomirVarga** [comment]
Hi, I have tried to follow
https://github.com/jrudolph/sbt-dependency-graph/tree/0.8 "how to use"
part, but I am not able to run dependency analyze. I have removed from
~/.sbt/0.13/global.sbt file what been there from 0.7.5 version and I have
updated ~/.sbt/0.13/plugins/plugins.sbt file. Current try-out will fail
like this:

sbt dependency-tree
[info] Loading global plugins from /home/luvar/.sbt/0.13/plugins
[info] Loading project definition from /home/user/workspace/secret/project
[info] Set current project to cloudfarms (in build
file:/home/user/workspace/secret/)
[error] Not a valid command: dependency-tree (similar: dependencies)
[error] Not a valid project ID: dependency-tree
[error] Expected ':' (if selecting a configuration)
[error] Not a valid key: dependency-tree (similar: dependency-overrides,
dependencyOverrides, dependency-classpath)
[error] dependency-tree
[error]                ^

What should I try?

On Wed, Nov 18, 2015 at 6:14 PM, Johannes Rudolph <notifications@github.com>
wrote:

> I released a new version 0.8.0-beta1 (to sonatype, wait a few hours to let
> it appear on maven central) which should have fixed this issue.
>
> @LubomirVarga <https://github.com/LubomirVarga>, could you check if this
> has fixed your problem?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jrudolph/sbt-dependency-graph/issues/77#issuecomment-157783331>
> .
>



-- 
Lubomir Varga

**Comment by jrudolph** [comment]
@LubomirVarga Are you using sbt >= 0.13.8?

**Comment by LubomirVarga** [comment]
I hope yes. End of sbt about output is:

[info] This is sbt 0.13.8
[info] The current project is {file:/home/user/workspace/secret/}root
1.1.178
[info] The current project is built against Scala 2.11.7
[info] Available Plugins: sbt.plugins.IvyPlugin, sbt.plugins.JvmPlugin,
sbt.plugins.CorePlugin, sbt.plugins.JUnitXmlReportPlugin,
net.virtualvoid.sbt.graph.DependencyGraphPlugin, play.Play, play.PlayJava,
play.PlayScala, play.twirl.sbt.SbtTwirl, com.typesafe.sbt.jse.SbtJsEngine,
com.typesafe.sbt.jse.SbtJsTask, com.typesafe.sbt.web.SbtWeb,
com.typesafe.sbt.webdriver.SbtWebDriver, com.typesafe.sbt.jshint.SbtJSHint,
com.typesafe.sbt.rjs.SbtRjs, com.typesafe.sbt.digest.SbtDigest,
com.typesafe.sbt.mocha.SbtMocha, com.typesafe.sbt.gzip.SbtGzip,
sass.SbtSass, org.sbtidea.SbtIdeaPlugin, com.typesafe.sbt.SbtEchoPlay,
com.typesafe.sbt.SbtEcho, com.typesafe.sbteclipse.plugin.EclipsePlugin,
com.typesafe.sbt.SbtNativePackager
[info] sbt, sbt plugins, and build definitions are using Scala 2.10.4

Also all submodules in project uses 0.13.8 sbt:
sbt sbtVersion
[info] Loading global plugins from /home/user/.sbt/0.13/plugins
[info] Loading project definition from /home/user/workspace/secret/project
[info] Set current project to secret (in build
file:/home/luvar/workspace/secret/)
[info] sync/*:sbtVersion
[info]  0.13.8
[info] data/*:sbtVersion
[info]  0.13.8
[info] mobileApi/*:sbtVersion
[info]  0.13.8
[info] root/*:sbtVersion
[info]  0.13.8

Lubo

On Thu, Nov 19, 2015 at 8:24 AM, Johannes Rudolph <notifications@github.com>
wrote:

> @LubomirVarga <https://github.com/LubomirVarga> Are you using sbt >=
> 0.13.8?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jrudolph/sbt-dependency-graph/issues/77#issuecomment-157975346>
> .
>



-- 
Lubomir Varga

**Comment by jrudolph** [comment]
Can you also try `sbt plugins`?

**Comment by LubomirVarga** [comment]
Yes, sure. Here it is:

sbt plugins
[info] Loading global plugins from /home/user/.sbt/0.13/plugins
[info] Loading project definition from /home/user/workspace/secret/project
[info] Set current project to secret (in build
file:/home/user/workspace/secret/)
In file:/home/user/workspace/secret/
        sbt.plugins.IvyPlugin: enabled in root, data,
cloudfarms-integrator-common, sync, mobileApi
        sbt.plugins.JvmPlugin: enabled in root, data,
cloudfarms-integrator-common, sync, mobileApi
        sbt.plugins.CorePlugin: enabled in root, data,
cloudfarms-integrator-common, sync, mobileApi
        sbt.plugins.JUnitXmlReportPlugin: enabled in root, data,
cloudfarms-integrator-common, sync, mobileApi
        net.virtualvoid.sbt.graph.DependencyGraphPlugin: enabled in root,
data, cloudfarms-integrator-common, sync, mobileApi
        play.Play: enabled in root, data, sync, mobileApi
        play.PlayJava: enabled in root, data, sync, mobileApi
        play.PlayScala
        play.twirl.sbt.SbtTwirl: enabled in root, data, sync, mobileApi
        com.typesafe.sbt.jse.SbtJsEngine: enabled in root, data, sync,
mobileApi
        com.typesafe.sbt.jse.SbtJsTask: enabled in root, data, sync,
mobileApi
        com.typesafe.sbt.web.SbtWeb: enabled in root, data, sync, mobileApi
        com.typesafe.sbt.webdriver.SbtWebDriver: enabled in root, data,
sync, mobileApi
        com.typesafe.sbt.jshint.SbtJSHint: enabled in root, data, sync,
mobileApi
        com.typesafe.sbt.rjs.SbtRjs: enabled in root, data, sync, mobileApi
        com.typesafe.sbt.digest.SbtDigest: enabled in root, data, sync,
mobileApi
        com.typesafe.sbt.mocha.SbtMocha: enabled in root, data, sync,
mobileApi
        com.typesafe.sbt.gzip.SbtGzip: enabled in root, data, sync,
mobileApi
        sass.SbtSass: enabled in root, data, sync, mobileApi


On Thu, Nov 19, 2015 at 9:52 AM, Johannes Rudolph <notifications@github.com>
wrote:

> Can you also try sbt plugins?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jrudolph/sbt-dependency-graph/issues/77#issuecomment-157992382>
> .
>



-- 
Lubomir Varga

**Comment by jrudolph** [comment]
Strange, everything seems fine. It's hard to say if this is something in your environment or another problem. Have you tried it on another project?

**Comment by jrudolph** [comment]
Other things you could try:

 * `sbt compile:dependencyTree`
 * `sbt "inspect dependencyTree"`

**Comment by LubomirVarga** [comment]
Yes, now I have.
I have tried it on one small submodule and also on one another (smaller)
project, where 0.7.5 version does not trigger stack overflow problem. On
given project it failed in same manner. On small submodule it have worked
nicely. Ehm...

I have not found possible cause of problem. I have copied given smaller
project to /tmp directory and it worked there. (sbt dependency-tree). Than
I have started new console and opened given project again in original place
(there are no spaces in path :). It failed again with non valid command. I
am confused.

I have repeated same thing with bigger project and it have also worked.
Strange.

So I can confirm that this problem is fixed in beta version now. Thanks,
Lubo

On Thu, Nov 19, 2015 at 10:33 AM, Johannes Rudolph <notifications@github.com
> wrote:

> Strange, everything seems fine. It's hard to say if this is something in
> your environment or another problem. Have you tried it on another project?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jrudolph/sbt-dependency-graph/issues/77#issuecomment-158002903>
> .
>



-- 
Lubomir Varga
